src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -1010,19 +1130,42 @@ struct BlockZipperJoinImpl {
         blockStatus = BlockStatus::allFilled;
       }
     };
+
     // We are only guaranteed to have all relevant blocks from one side, so we
     // also need to pass through the remaining blocks from the other side.
     while (!equalToCurrentElLeft.empty() && !equalToCurrentElRight.empty()) {
+      switch (blockStatus) {
+        case BlockStatus::allFilled:
+          addCartesianProduct(equalToCurrentElLeft, rightSide_.undefBlocks_);
+          addCartesianProduct(leftSide_.undefBlocks_, equalToCurrentElRight);
+          break;
+        case BlockStatus::rightMissing:
+          addCartesianProduct(leftSide_.undefBlocks_, equalToCurrentElRight);
+          break;
+        case BlockStatus::leftMissing:
+          addCartesianProduct(equalToCurrentElLeft, rightSide_.undefBlocks_);
+          break;
+        default:
+          AD_FAIL();
+      }
       addAll<DoOptionalJoin>(equalToCurrentElLeft, equalToCurrentElRight);
       switch (blockStatus) {
         case BlockStatus::allFilled:
+          storeNewlyFoundUndefBlocks(equalToCurrentElRight,
+                                     rightSide_.undefBlocks_);
+          storeNewlyFoundUndefBlocks(equalToCurrentElLeft,
```

Comment:
How much simpler does the code get , if you
1. right at the beginning, as a first step, materialize and store all Undef blocks from Left and Right, write their Cartesian product to the result.
2. Then you always know "I have seen all undef" as well as "my actual new input never contains additional undef".
3. You would probably have to copy/clone a partial block, (the one per side where there first are UNDEF values, and then Non-Undef values). 

A) 
Maybe even do this completely outside the class (with additional generator wrappiings etc), s.t. your impl directly gets
for each side the "undefblocks" as well as the "remainder which is guaranteed not to be UNDEF".